### ISSUE TITLE
Is there any support for visualisation like precision, recall, model_visualisation, training_loss, validation_losss? 

### ISSUE THREAD

**Issue by DL8614** [muhanzhang/pytorch_DGCNN#26]
Hi,
My dataset consists of 20 classes and I will also add more class in future. Node contains only label integer as features. So is it still possible to classify whole graph? I think it should work because for sort pooling layer, it doesn#t matter whether nodes have label fetaures or continous features.
Am I correct?
and can you give also some support for visualisation of network?

**Comment by muhanzhang** [comment]
Yes, it is possible to classify whole graph. Integer initial node labels are absolutely fine. As for visualization, you may refer to Figure 2 of [paper](https://www.cse.wustl.edu/~muhan/papers/AAAI_2018_DGCNN.pdf).

**Comment by DL8614** [comment]
Hi, Thanks for fast reply.
In ReadMe for own Dataset, what do you mean by Node tag? Is it node label(integer)?

**Comment by muhanzhang** [comment]
Yes, node tag is integer node label.

**Comment by DL8614** [comment]
Hi,
Do you have experience about using this pipeline for object recognition. I mean one object can be also represented with help of many features like shape, color, text. And If I have this feature as a label in graph format, may I classify whole object in one specific class?

**Comment by muhanzhang** [comment]
Does the object have clear graph structures? Such as connections between its different components where each component has properties (features or labels). If so, it is a graph classification problem and the software might help.

**Comment by DL8614** [comment]
Yes, I have lables for each component features. Then I will try my data on this pipeline. One more question is that, what about if I have two objects A & B in one image. And in training I have trained the model with Graph of A and B both separately. And then If test image have A and B together, can network cut one test graph into two graph and classify them in two separate class?

**Comment by muhanzhang** [comment]
If I understand correctly, this is a multi-label classification problem. You may alternatively optimize a multi-label loss (e.g., sum of cross entropy losses for all labels), and then let the model output the prediction (yes or no) for each label. Otherwise, the current model optimizes a multi-class loss can only output one predicted class.

**Comment by DL8614** [comment]
What about, if I use fully connected Graph or Dense Graph as an input? Is it allowed to this frameowrk? And one more question is that, how can I understand, structural role of node in Graph? This comes in my questions, when I was reading about sortpooling layer.

**Comment by muhanzhang** [comment]
Fully connected or dense graph is allowed. But there graph structures will not make sense any more in classification and some statistics like node number/feature distribution might play the classifiation role. By structural role of node we usually mean how a node is positioned relatively in a graph; e.g., a node centered in a graph with many neighbors has a different structural role from a node locating in the margin; many nodes can share symmetric structural roles too. You may refer to my Weisfeiler-Lehman Neural Machine paper for more details. Thanks.

**Comment by DL8614** [comment]
In my dataset, I have number of node labels are around 300. And it is too many in comarision to other dataset like MUTAG, ENZYMES. Do you have any idea, how it affects the learning process. Or should I produce more training data as increasing the number of node labels?

**Comment by muhanzhang** [comment]
It probably won't have huge influence on the learning, since we can just treat their one-hot encodings as 300-dimension node features. Nevertheless, it might require more training data and more optimization epochs, etc. A workaround is to learn low-dimensional embeddings for node labels using a look-up table. You may just insert this look-up table before the graph convolution and use the look-up table to extract node features.

**Comment by DL8614** [comment]
What do you mean by a look-up table? Should I do create separately one look-up table or what? 
One more question is that, which kind of graph format has all other datasets like MUTAG, ENZYMES?
In my case, I have undirected edges and I prepared data according to your suggestion in text format as well as stored in Matlab file.

**Comment by muhanzhang** [comment]
For lookup table, please refer to https://pytorch.org/docs/stable/nn.html#embedding for more details. It is mainly used to map each discrete label to a low-dimensional embedding vector before inputing a some neural networks, like word embeddings. For graph format, please use the txt format. The matlab format is traditionally used in graph kernel methods. You may use mat2txt.m to transform them to .txt format.
